File: GFG POTD/020-July-2025/007-next-greater-element.cpp
// https://www.geeksforgeeks.org/problems/next-greater-element/1




#include <bits/stdc++.h>
using namespace std;


class Solution {
  private:
    // TC : O(N)
    // SC : O(N)
    vector<int> approach_1(vector<int> &arr) {
        int n = arr.size();
        for (int i = 0; i < n; i ++) {
            arr.push_back(arr[i]);
        }

        stack<int> stk;
        vector<int> result(n, -1);
        int len = arr.size(); 
        
        for (int i = len - 1; i >= 0; i --) {
            while (!stk.empty() && arr[stk.top()] <= arr[i]) {
                stk.pop();
            }
            if (!stk.empty()) result[i % n] = arr[stk.top()];
            stk.push(i); 
        }
        return result;
    }
    
    // TC : O(N)
    // SC : O(N)
    vector<int> approach_2(vector<int> &arr) {
        int n = arr.size();
        
        stack<int> stk;
        vector<int> result(n, -1);
        int len = 2 * arr.size(); 
        
        for (int i = len - 1; i >= 0; i --) {
            while (!stk.empty() && arr[stk.top() % n] <= arr[i % n]) {
                stk.pop();
            }
            if (!stk.empty()) result[i % n] = arr[stk.top() % n];
            stk.push(i); 
        }
        return result;
    }
  public:
    vector<int> nextLargerElement(vector<int> &arr) {
        // return approach_1(arr);
        return approach_2(arr);  // No input modify
    }
};

File: GFG POTD/020-July-2025/007-next-greater-element.java
// https://www.geeksforgeeks.org/problems/next-greater-element/1




import java.util.*;

class Solution {
    public ArrayList<Integer> nextLargerElement(int[] arr) {
        int n = arr.length;
        
        Stack<Integer> stk = new Stack<>();
        ArrayList<Integer> result = new ArrayList<>(Collections.nCopies(n, -1));
        int len = 2 * n;
        
        for (int i = len - 1; i >= 0; i --) {
            while (!stk.isEmpty() && arr[stk.peek() % n] <= arr[i % n]) {
                stk.pop();
            }
            if (!stk.isEmpty()) result.set(i % n, arr[stk.peek() % n]);
            stk.add(i);
        }
        
        return result;
    }
}